stone/target/js_types.py

```diff
@@ -0,0 +1,211 @@
+                    self._generate_type(data_type, extra_args.get(data_type, []))
+
+    def _parse_extra_args(self, api, extra_args_raw):
+        """
+        Parses extra arguments into a map keyed on particular data types.
+        """
+        extra_args = {}
+
+        for extra_arg_raw in extra_args_raw:
+            def exit(m):
+                print('Invalid --extra-arg:%s: %s' % (m, extra_arg_raw),
+                      file=sys.stderr)
+                sys.exit(1)
+
+            try:
+                extra_arg = json.loads(extra_arg_raw)
+            except ValueError as e:
+                exit(str(e))
+
+            # Validate extra_arg JSON blob
+            if 'match' not in extra_arg:
+                exit('No match key')
+            elif (not isinstance(extra_arg['match'], list) or
+                      len(extra_arg['match']) != 2):
+                exit('match key is not a list of two strings')
+            elif (not isinstance(extra_arg['match'][0], six.text_type) or
+                      not isinstance(extra_arg['match'][1], six.text_type)):
+                print(type(extra_arg['match'][0]))
+                exit('match values are not strings')
+            elif 'arg_name' not in extra_arg:
+                exit('No arg_name key')
+            elif not isinstance(extra_arg['arg_name'], six.text_type):
+                exit('arg_name is not a string')
+            elif 'arg_type' not in extra_arg:
+                exit('No arg_type key')
+            elif not isinstance(extra_arg['arg_type'], six.text_type):
+                exit('arg_type is not a string')
+            elif ('arg_docstring' in extra_arg and
+                      not isinstance(extra_arg['arg_docstring'], six.text_type)):
+                exit('arg_docstring is not a string')
+
+            attr_key, attr_val = extra_arg['match'][0], extra_arg['match'][1]
+
+            extra_args.setdefault(attr_key, {})[attr_val] = \
+                (extra_arg['arg_name'], extra_arg['arg_type'],
+                 extra_arg.get('arg_docstring'))
+
+        # Extra arguments, keyed on data type objects.
+        extra_args_for_types = {}
+        # Locate data types that contain extra arguments
+        for namespace in api.namespaces.values():
+            for route in namespace.routes:
+                extra_parameters = []
+                if is_user_defined_type(route.arg_data_type):
+                    for attr_key in route.attrs:
+                        if attr_key not in extra_args:
+                            continue
+                        attr_val = route.attrs[attr_key]
+                        if attr_val in extra_args[attr_key]:
+                            extra_parameters.append(extra_args[attr_key][attr_val])
+                if len(extra_parameters) > 0:
+                    extra_args_for_types[route.arg_data_type] = extra_parameters
+
+        return extra_args_for_types
+
+    def _generate_type(self, data_type, extra_parameters = []):
```

Comment:
1. No spaces around the `=`.
2. Standard practice is to avoid using a list as a default argument. You can read why here: http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments